monai/transforms/croppad/dictionary.py

```diff
@@ -536,7 +536,7 @@ def __call__(self, data: Mapping[Hashable, np.ndarray]) -> List[Dict[Hashable, n
         ret = []
         d = dict(data)
         for i in range(self.num_samples):
-            cropped = self.cropper(d)
+            cropped = self.cropper(deepcopy(d))
```

Comment:
Hi @rijobro ,

I think we should try to avoid copy the whole data as some input images may be very big.
Could you please change to only deepcopy some data like:
https://github.com/Project-MONAI/MONAI/blob/master/monai/transforms/croppad/dictionary.py#L805
@ericspod @wyli  what do you think?

Thanks.

Reply:
Do we need deepcopy here at all? I don't know if there's a specific issue with not copying images and just do a shallow copy of the dictionary.

Reply:
Hi @ericspod ,

Shallow copy will share the `dict` object in it.
In this PR, the `d = dict(data)` will share same `img_transforms` dict in it, then this line put transform information into the same dict.:
https://github.com/Project-MONAI/MONAI/blob/master/monai/transforms/croppad/dictionary.py#L439

Thanks.

Reply:
@ericspod at least **some** of the dictionary needs to be deep copied in order to fix the linked issue. Howver, @Nic-Ma, I would argue that the whole dictionary needs to be deep copied. Without this PR, the following code:

```python
data = {"img": np.ones((1, 10, 11, 12)), "label": np.ones((1, 10, 11, 12))}

sampler = RandSpatialCropSamplesd(
    keys=["img"],
    roi_size=(3, 3, 3),
    num_samples=2,
    random_center=True,
    random_size=False,
)
samples = sampler(data)
# print label 1 max
print(samples[1]["label"].max())
# change label 0
samples[0]["label"] *= 2
# re-print label 1 max
print(samples[1]["label"].max())
```
gives the output:
```
1.0
2.0
```
whereas with the PR, we get:
```
1.0
1.0
```
I think the latter is desirable.

Reply:
Hi @rijobro ,

Sorry maybe I didn't make it clear, I mean in order to save memory, we can deepcopy the items not in `keys`, like:
https://github.com/Project-MONAI/MONAI/blob/master/monai/transforms/croppad/dictionary.py#L805
What do you think?

Thanks.

Reply:
Hi @rijobro , do you want to update this logic referring to the `RandCropByPosNegRatio`?
Thanks.

Reply:
I submitted commit to update this logic as @rijobro  is PTO.
Thanks.